File: stresstest/worker.mjs
import { getClient } from 'leased-clickhouse';

async function run(id) {
    console.log(`Worker ${id} starting...`);
    const db = await getClient({ dataDir: './stress-db' });
    await db.exec({
        query: 'CREATE TABLE IF NOT EXISTS workers (id UInt32, val String) ENGINE = MergeTree ORDER BY id'
    });
    await db.insert({
        table: 'workers',
        values: [{ id: parseInt(id), val: `worker-${id}` }],
        format: 'JSONEachRow'
    });
    const result = await db.query({ query: 'SELECT count() as c FROM workers' });
    const json = await result.json();
    console.log(`Worker ${id} sees ${json.data[0].c} rows.`);
}

run(process.argv[2] || '0').catch(console.error);

File: py-stress/worker.py
"""A worker process that inserts data and occasionally crashes."""
import sys
import time
import random
import clickhouse_sidecar

def run(wid):
    client = clickhouse_sidecar.get_client("stress-py-db")
    client.command("CREATE TABLE IF NOT EXISTS stress_events (id UInt32, wid UInt32) ENGINE = MergeTree ORDER BY id")

    # 20% chance to crash mid-execution without cleanly closing the client
    if random.random() < 0.2:
        print(f"Worker {wid} crashing intentionally!")
        sys.exit(1)

    for i in range(10):
        client.insert("stress_events", [[int(time.time() * 1000) % 1000000 + i, int(wid)]], column_names=["id", "wid"])
        time.sleep(0.05)

    result = client.query(f"SELECT count() FROM stress_events WHERE wid = {wid}")
    print(f"Worker {wid} sees {result.result_rows[0][0]} of its own rows.")
    client.close()

if __name__ == "__main__":
    run(sys.argv[1] if len(sys.argv) > 1 else "0")
